### PR TITLE
Kubectl `kcp workspace` plugin

### PR DESCRIPTION
**Pull request by davidfestal**
Adds a `kcp workspace` `kubectl` plugin to manage the workspace through the `workspaces` virtual workspace.

This also adds a new demo showing how to use the kubectl plugin, as well as the required setup to have it working.

### CODE REVIEW

File: contrib/demo/ingress-script/ingress [hunk 1]
```diff
@@ -1,10 +1,12 @@
 #!/usr/bin/env bash
 
 DEMO_DIR="$( dirname "${BASH_SOURCE[0]}" )"
-. ${DEMO_DIR}/demo-magic
+DEMOS_DIR="$(cd "${DEMO_DIR}/.."; pwd)"
```
Comment (stevekuznetsov) [inline contrib/demo/ingress-script/ingress:4]: Use `realpath`

File: contrib/demo/ingress-script/ingress.result [hunk 2]
```diff
@@ -9,7 +9,7 @@ cluster.cluster.example.dev/us-west1 condition met
 ingress-demo $ [1;37mkubectl apply -f /home/jmprusi/go/src/github.com/jmprusi/kcp/contrib/demo/clusters/kind/us-east1.yaml[0m
 cluster.cluster.example.dev/us-east1 created
 ingress-demo $ [1;37mkubectl wait cluster/us-east1 --for condition=Ready --timeout=60s[0m
-cluster.cluster.example.dev/us-east1 condition met
+cluster.cluster.example.dev/us-east1 condition met../
```
Comment (stevekuznetsov) [inline contrib/demo/ingress-script/ingress.result:12]: ?

File: contrib/demo/runDemoScripts.sh [hunk 3]
```diff
@@ -15,8 +15,8 @@
 # limitations under the License.
 
 CURRENT_DIR="$(pwd)"
-DEMO_ROOT="$(cd $(dirname "${BASH_SOURCE}") && pwd)"
-KCP_ROOT="$(cd ${DEMO_ROOT}/../.. && pwd)"
+DEMOS_ROOT="$(cd $(dirname "${BASH_SOURCE}") && pwd)"
+KCP_ROOT="$(cd ${DEMOS_ROOT}/../.. && pwd)"
```
Comment (stevekuznetsov) [inline contrib/demo/runDemoScripts.sh:19]: Can you break this into a separate PR? Seems unrelated to the rest of the content in here

File: contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+DEMO_DIR="$( dirname "${BASH_SOURCE[0]}" )"
+DEMOS_DIR="$(cd "${DEMO_DIR}/.."; pwd)"
+
+. ${DEMOS_DIR}/demo-magic
```
Comment (stevekuznetsov) [inline contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin:20]: nit: `source` is easier to read

File: contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+
+ROOT_DIR="$( cd ${DEMOS_DIR}/../.. && pwd)"
```
Comment (stevekuznetsov) [inline contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin:22]: If all of this preamble is shared with others, why not put it into a shared file you `source`?

File: contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+pe "echo \${KUBECONFIG}"
+pe "more ${KUBECONFIG}"
+pe ""
+
+pause
+clear
+
+# creates the cluster roles + bindings at the workspace level + workspace content level.
+
+pe "kubectl kcp workspace --token user-1-token create workspace11"
```
Comment (stevekuznetsov) [inline contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin:71]: This demo was really hard for me to follow. Can we do some or all of the following:

 - prefix workspace names with their users, so `workspace11` becomes `user-1-workspace-1`
 - not mix current contexts that are user-specific with other tokens

Especially the second point - it's very opaque what is happening here. I believe with OIDC you *only* authenticate with the token, right? But the workspace context is user-specific? It's not clear.

File: contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin-prepareClusters.sh [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
```
Comment (stevekuznetsov) [inline contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin-prepareClusters.sh:1]: Instead of checking in empty files, just don't call this unless it exists?

File: contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin-removeClusters.sh [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env bash
```
Comment (stevekuznetsov) [inline contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin-removeClusters.sh:1]: Instead of checking in empty files, just don't call this unless it exists?

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 9]
```diff
@@ -0,0 +1,570 @@
...
+	cmd.AddCommand(deleteCmd)
+	return cmd, nil
+}
+
+func (o *WorkspaceOptions) Complete(cmd *cobra.Command) error {
+	if err := metav1.AddMetaToScheme(scheme.Scheme); err != nil {
+		return err
+	}
+
+	kubectlConfigOverrideFlags := clientcmd.RecommendedConfigOverrideFlags("")
```
Comment (stevekuznetsov) [inline pkg/cliplugins/workspace/cmd/workspace.go:215]: What is going on here?

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 10]
```diff
@@ -0,0 +1,570 @@
...
+}
+
+func PrioritizedAuthInfo(values ...*api.AuthInfo) *api.AuthInfo {
+	for _, value := range values {
+		if value == nil {
+			continue
+		}
+		value := *value
+		if value.Token != "" || value.TokenFile != "" || value.Password != "" || value.Username != "" {
+			return &value
```
Comment (stevekuznetsov) [inline pkg/cliplugins/workspace/cmd/workspace.go:311]: Why are we doing a shallow copy in here?

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 11]
```diff
@@ -0,0 +1,570 @@
...
+
+	overrides := &clientcmd.ConfigOverrides{
+		AuthInfo:       *PrioritizedAuthInfo(&o.workspaceDirectoryOverrides.AuthInfo, &o.kubectlOverrides.AuthInfo, currentContextAuthInfo),
+		ClusterInfo:    o.workspaceDirectoryOverrides.ClusterInfo,
+		CurrentContext: o.workspaceDirectoryOverrides.CurrentContext,
+	}
+	return clientcmd.NewDefaultClientConfig(*o.startingConfig, overrides).ClientConfig()
+}
+
+func getScopeAndName(workspaceKey string) (string, string) {
```
Comment (stevekuznetsov) [inline pkg/cliplugins/workspace/cmd/workspace.go:336]: This is a pretty heavy-weight file. Do we want any part of it unit-tested?

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 12]
```diff
@@ -0,0 +1,570 @@
...
+	if !strings.HasPrefix(currentContextName, kcpWorkspaceContextNamePrefix) {
+		return "", "", errors.New("The current context is not a KCP workspace !")
+	}
+
+	scope, workspaceName := getScopeAndName(strings.TrimPrefix(currentContextName, kcpWorkspaceContextNamePrefix))
+	return scope, workspaceName, nil
+}
+
+func (o *WorkspaceOptions) checkCurrentWorkspace(workspaceName string, tenancyClient *tenancyclient.TenancyV1alpha1Client) (err error) {
+	if _, err := tenancyClient.Workspaces().Get(context.TODO(), workspaceName, metav1.GetOptions{}); err != nil {
```
Comment (stevekuznetsov) [inline pkg/cliplugins/workspace/cmd/workspace.go:432]: No `context.TODO()` should merge IMO - please pass down an appropriate context.

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 13]
```diff
@@ -0,0 +1,570 @@
...
+		return err
```
Comment (stevekuznetsov) [inline pkg/cliplugins/workspace/cmd/workspace.go:433]: Anywhere you do:

```go
if err := thing(); err != nil {
    return err
}
return nil
```

You can just:

```go
return thing()
```

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 14]
```diff
@@ -0,0 +1,570 @@
...
+		return nil
+	}
+
+	if err != nil {
+		return err
+	}
+
+	// Check that the scope is consistent with the workspace-directory config
+	if !strings.HasSuffix(workspaceDirectoryRestConfig.Host, "/"+scope) {
+		_ = outputCurrentWorkspaceMessage()
```
Comment (stevekuznetsov) [inline pkg/cliplugins/workspace/cmd/workspace.go:465]: Is this a side-effect call?

File: pkg/virtual/workspaces/registry/kubeconfig_rest_test.go [hunk 15]
```diff
@@ -807,7 +807,7 @@ func TestKubeconfigFailBecauseShardNotFound(t *testing.T) {
 	test := TestDescription{
 		TestData: TestData{
 			user:  user,
-			scope: organizationScope,
+			scope: OrganizationScope,
```
Comment (stevekuznetsov) [inline pkg/virtual/workspaces/registry/kubeconfig_rest_test.go:810]: IMHO test fixtures should explicitly not use constants and helpers - if we are bound to an outward-facing API we need to know that a client's expected behavior will be broken if (for instance) I change this constant incorrectly.

File: cmd/kubectl-kcp-workspace/main.go [hunk 16]
```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2021 The KCP Authors.
```
Comment (sttts) [inline cmd/kubectl-kcp-workspace/main.go:2]: nit: 22

File: go.mod [hunk 17]
```diff
@@ -25,7 +24,8 @@ require (
 	k8s.io/apiextensions-apiserver v0.0.0
 	k8s.io/apimachinery v0.0.0
 	k8s.io/apiserver v0.0.0
-	k8s.io/client-go v0.0.0
+	k8s.io/cli-runtime v0.23.1
+	k8s.io/client-go v0.23.1
```
Comment (sttts) [inline go.mod:28]: why does go mod add tags here? Not wrong, replaced anyway, just curious.

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 18]
```diff
@@ -0,0 +1,570 @@
...
+	cmd.AddCommand(useCmd)
+	cmd.AddCommand(currentCmd)
+	cmd.AddCommand(listCmd)
+	cmd.AddCommand(createCmd)
+	cmd.AddCommand(deleteCmd)
+	return cmd, nil
+}
+
+func (o *WorkspaceOptions) Complete(cmd *cobra.Command) error {
+	if err := metav1.AddMetaToScheme(scheme.Scheme); err != nil {
```
Comment (sttts) [inline pkg/cliplugins/workspace/cmd/workspace.go:211]: why this? Modifying somebody else' scheme is not good

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 19]
```diff
@@ -0,0 +1,570 @@
...
+		}
+		o.startingConfig.Clusters[o.workspaceDirectoryOverrides.CurrentContext] = workspaceDirectoryCluster
+		o.startingConfig.Contexts[o.workspaceDirectoryOverrides.CurrentContext] = workspaceDirectoryContext
+
+		return clientcmd.ModifyConfig(o.configAccess, *o.startingConfig, true)
+	}
+	return nil
+}
+
+func PrioritizedAuthInfo(values ...*api.AuthInfo) *api.AuthInfo {
```
Comment (sttts) [inline pkg/cliplugins/workspace/cmd/workspace.go:304]: needs go doc.

Why is this a public func? Used elsewhere?

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 20]
```diff
@@ -0,0 +1,570 @@
...
+
+const (
+	kcpWorkspaceContextNamePrefix  string = "workspace.kcp.dev/"
+	kcpPreviousWorkspaceContextKey string = "workspace.kcp.dev/-"
+)
+
+// WorkspaceOptions provides information required to update
+// the current context on a user's KUBECONFIG
+type WorkspaceOptions struct {
+	workspaceDirectoryOverrides *clientcmd.ConfigOverrides
```
Comment (sttts) [inline pkg/cliplugins/workspace/cmd/workspace.go:64]: what is a directory?

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 11 above)
Comment (sttts) [reply pkg/cliplugins/workspace/cmd/workspace.go:336]: and security sensitive.

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 21]
```diff
@@ -0,0 +1,570 @@
...
+
+	if !strings.HasPrefix(currentContextName, kcpWorkspaceContextNamePrefix) {
+		return "", "", errors.New("The current context is not a KCP workspace !")
+	}
+
+	scope, workspaceName := getScopeAndName(strings.TrimPrefix(currentContextName, kcpWorkspaceContextNamePrefix))
+	return scope, workspaceName, nil
+}
+
+func (o *WorkspaceOptions) checkCurrentWorkspace(workspaceName string, tenancyClient *tenancyclient.TenancyV1alpha1Client) (err error) {
```
Comment (sttts) [inline pkg/cliplugins/workspace/cmd/workspace.go:431]: what does check mean? Need go doc and/or a better name.

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 11 above)
Comment (sttts) [reply pkg/cliplugins/workspace/cmd/workspace.go:336]: also needs go doc

File: pkg/cliplugins/workspace/cmd/workspace.go [hunk 22]
```diff
@@ -0,0 +1,570 @@
...
+	}
+	return nil
+}
+
+// Validate ensures that all required arguments and flag values are provided
+func (o *WorkspaceOptions) Validate() error {
+	return nil
+}
+
+func (o *WorkspaceOptions) ensureWorkspaceDirectoryContextExists() error {
```
Comment (sttts) [inline pkg/cliplugins/workspace/cmd/workspace.go:269]: could this be functional without side-effects? Would be clearer. This modifies o heavily it seems.

File: contrib/demo/ingress-script/ingress (hunk 1 above)
Comment (davidfestal) [reply contrib/demo/ingress-script/ingress:4]: afaict `realpath` doesn't exist on MacOS

File: contrib/demo/runDemoScripts.sh (hunk 3 above)
Comment (davidfestal) [reply contrib/demo/runDemoScripts.sh:19]: Done in PR #381

File: contrib/demo/ingress-script/ingress.result (hunk 2 above)
Comment (davidfestal) [reply contrib/demo/ingress-script/ingress.result:12]: Fixed in PR #381

File: cmd/kubectl-kcp-workspace/main.go (hunk 16 above)
Comment (marun) [reply cmd/kubectl-kcp-workspace/main.go:2]: - Does the year being correct only matter for newly added files? 
- Is there any consequence to it being wrong?

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 18 above)
Comment (davidfestal) [reply pkg/cliplugins/workspace/cmd/workspace.go:211]: The `TableXXX` types that are in the meta v1 scheme are not added by default to the generated KCP clientset scheme. So that when I want to get the result of a request done with this clientset as a table, it doesn't know the `Table` types.
It happens that in the `kubectl` code, those types are added to the main `kubectl` scheme.

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 19 above)
Comment (davidfestal) [reply pkg/cliplugins/workspace/cmd/workspace.go:304]: Nowhere you're right. I'll change that.

File: go.mod (hunk 17 above)
Comment (davidfestal) [reply go.mod:28]: I brought the v0.0.0 back and it works well. Probably VSCode overdoing it.

File: cmd/kubectl-kcp-workspace/main.go (hunk 16 above)
Comment (davidfestal) [reply cmd/kubectl-kcp-workspace/main.go:2]: Will fix

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 9 above)
Comment (davidfestal) [reply pkg/cliplugins/workspace/cmd/workspace.go:215]: I'll explain in comments

File: contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin-removeClusters.sh (hunk 8 above)
Comment (davidfestal) [reply contrib/demo/workspaceKubectlPlugin-script/workspaceKubectlPlugin-removeClusters.sh:1]: Will do. Required change has been done in PR #381

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 21 above)
Comment (davidfestal) [reply pkg/cliplugins/workspace/cmd/workspace.go:431]: Will change the name to `checkWorkspaceExists`

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 18 above)
Comment (davidfestal) [reply pkg/cliplugins/workspace/cmd/workspace.go:211]: Any suggestion to do better ?

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 14 above)
Comment (davidfestal) [reply pkg/cliplugins/workspace/cmd/workspace.go:465]: No, it simply writes to the output of the CLI.

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 13 above)
Comment (davidfestal) [reply pkg/cliplugins/workspace/cmd/workspace.go:433]: Sure.

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 22 above)
Comment (davidfestal) [reply pkg/cliplugins/workspace/cmd/workspace.go]: Nor really, since the idea is to add a context for the *workspace-directory* (== the `workspaces` virtual workspace) in the current Kubeconfig if none exists.
I commented the method to make it clear though.

**Comment by davidfestal** [comment]
@stevekuznetsov @sttts I think I address your comments, either by answering or by fixing it in the code. For this I also refactored the structure of the code to split it into several files, which I hope makes it clearer.

File: pkg/cliplugins/workspace/cmd/workspace.go (hunk 14 above)
Comment (stevekuznetsov) [reply pkg/cliplugins/workspace/cmd/workspace.go]: That's a side-effect :)

File: cmd/kubectl-kcp-workspace/main.go (hunk 16 above)
Comment (stevekuznetsov) [reply cmd/kubectl-kcp-workspace/main.go]: @marun I am not authoritative but my guesses:
- yes
- no

**Review by sttts — APPROVED** [review]
